base/random.jl

```diff
@@ -683,6 +685,39 @@ end
 rand{T}(rng::AbstractRNG, r::AbstractArray{T}, dims::Dims) = rand!(rng, Array{T}(dims), r)
 rand(rng::AbstractRNG, r::AbstractArray, dims::Int...) = rand(rng, r, dims)
 
+# rand from a string
+
+isvalid_unsafe(s::String, i) = !Base.is_valid_continuation(unsafe_load(pointer(s), i))
+
+function rand(rng::AbstractRNG, s::String)::Char
+    g = RangeGenerator(1:s.len)
+    while true
+        pos = rand(rng, g)
+        isvalid_unsafe(s, pos) && return s[pos]
+    end
+end
+
+function rand(rng::AbstractRNG, s::AbstractString)::Char
+    g = RangeGenerator(1:endof(s))
+    while true
+        try # the generic `isvalid` includes an equivalent try/catch statement
+            return s[rand(rng, g)]
+        end
+    end
+end
```

Comment:
I think it would be better to do `i = rand(rng, g); isvalid(s, i) && return i` since various string types are expected to implement that without the try/catch. The try/catch only exists as a last resort fall-back, and I believe it is more targeted and only catches `UnicodeError` exceptions (if it's not already, it should be).